packages/editor-sdk/src/components/Widgets/ArrayField.tsx

```diff
@@ -36,81 +68,146 @@ export const ArrayField: React.FC<WidgetProps<ArrayFieldWidgetOptionsType>> = pr
-                }}
-              />
-              <IconButton
-                aria-label={`delete-${idx}`}
-                icon={<DeleteIcon />}
-                colorScheme="red"
-                onClick={() => {
-                  const newFormData = [...value];
-                  newFormData.splice(idx, 1);
-                  onChange(newFormData);
-                }}
-              />
-            </ButtonGroup>
-            <SchemaField
-              component={component}
-              schema={mergeWidgetOptionsIntoSchema(subSchema, { expressionOptions })}
-              value={v}
-              level={level + 1}
-              services={services}
-              onChange={newValue => {
-                const newFormData = [...value];
-                newFormData[idx] = newValue;
-                onChange(newFormData);
-              }}
-            />
-          </Box>
-        );
-      })}
-      <Flex justify="end">
+
+  const isNotBaseType = subSchema.type === 'object' || subSchema.type === 'array';
+  const itemInfos = value.map((v, idx) => {
+    const props = {
+      component,
+      schema: mergeWidgetOptionsIntoSchema(
+        {
+          ...subSchema,
+          title: isNotBaseType ? '' : subSchema.title,
+        },
+        {
+          expressionOptions,
+        }
+      ),
+      value: v,
+      path: path.concat(String(idx)),
+      level: level + 1,
+      services,
+      onChange: (newValue: any) => {
+        const newFormData = [...value];
+        newFormData[idx] = newValue;
+        onChange(newFormData);
+      },
+    };
+    const buttons = (
+      <ButtonGroup
+        spacing={0}
+        size="xs"
+        variant="ghost"
+        display="flex"
+        justifyContent="end"
+      >
         <IconButton
-          aria-label="add"
-          icon={<AddIcon />}
-          size="sm"
+          aria-label={`up-${idx}`}
+          icon={<ArrowUpIcon />}
+          disabled={idx === 0}
           onClick={() => {
-            onChange(value.concat(parseTypeBox(subSchema as TSchema)));
+            const newFormData = [...value];
+            swap(newFormData, idx, idx - 1);
+            onChange(newFormData);
           }}
         />
-      </Flex>
+        <IconButton
+          aria-label={`down-${idx}`}
+          icon={<ArrowDownIcon />}
+          disabled={idx === value.length - 1}
+          onClick={() => {
+            const newFormData = [...value];
+            swap(newFormData, idx, idx + 1);
+            onChange(newFormData);
+          }}
+        />
+        <IconButton
+          aria-label={`delete-${idx}`}
+          icon={<DeleteIcon />}
+          colorScheme="red"
+          onClick={() => {
+            const newFormData = [...value];
+            newFormData.splice(idx, 1);
+            onChange(newFormData);
+          }}
+        />
+      </ButtonGroup>
+    );
+
+    return {
+      props,
+      buttons,
+    };
+  });
+
+  return isNotBaseType ? (
```

Comment:
This return is too long and has too many ternary operator. Can we make this more readable?

Reply:
Or maybe we can split baseType array and non-baseType array to 2 components?

Reply:
Do you mean split it into two widgets? I think it is not good to do that because we should auto display the fittable content depending on different array types.

I have no good idea to simplify it now because the two type arrays need to reuse the buttons and sub props.

Reply:
I mean it can be one widget, while the implementation can be split to 2 or more components. This is for increacing the maintainability.

Reply:
Spliting code is not the only way to increase maintainability. Just removing the nested ternary operators also helps that.

Reply:
Updated.